### PR TITLE
Added Merge policy and merge planner.

### PR DESCRIPTION
**Pull request by fulmicoton**
Merge first stab.

It includes the merge policy logic and an actor receiving notification of new splits.

### CODE REVIEW

File: quickwit-indexing/Cargo.toml [hunk 1]
```diff
@@ -43,3 +43,4 @@ required-features = ["fail/failpoints"]
 quickwit-metastore = {path = "../quickwit-metastore", features=["testsuite"]}
 mockall = "0.9"
 quickwit-common = {path="../quickwit-common", version="0.1"}
+proptest = "1"
```
Comment (fmassot) [inline quickwit-indexing/Cargo.toml:46]: nice, didn't know proptest

File: quickwit-indexing/src/merge_policy.rs [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+            target_demux_ops: 1,
+            max_merge_docs: 10_000_000,
+            min_level_num_docs: 100_000,
+            merge_factor: 10,
+            merge_factor_max: 12,
+        }
+    }
+}
+
+fn remove_matching_els<T, Pred: Fn(&T) -> bool>(els: &mut Vec<T>, predicate: Pred) -> Vec<T> {
```
[the remark this answers is not in the document]
Comment (fmassot) [reply quickwit-indexing/src/merge_policy.rs:116]: elements?

File: quickwit-indexing/src/merge_policy.rs [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+            let time_range = split.time_range.as_ref().unwrap();
+            Reverse(*time_range.end())
+        });
+        // Splits should naturally have an increasing num_merge
+        let split_levels = self.build_split_levels(splits);
+        for split_range in split_levels.into_iter().rev() {
+            if split_range.len() < self.merge_factor {
+                continue;
+            }
+            // number of splits we will merge
```
Comment (fmassot) [inline quickwit-indexing/src/merge_policy.rs:167]: comment can be removed

File: quickwit-indexing/src/merge_policy.rs [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+                .sum::<usize>()
+                + splits.len(),
+            "The merge policy is supposed to keep the number of splits."
+        );
+        merge_operations
+    }
+
+    fn is_mature(&self, split: &SplitMetadata) -> bool {
+        // split.demux_signature == self.target_demux_ops &&
+        // Include demux signature in the is_mature logic.
```
Comment (fmassot) [inline quickwit-indexing/src/merge_policy.rs:195]:
```suggestion
        // TODO: include demux signature in the is_mature logic.
```

File: quickwit-indexing/src/merge_policy.rs [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+        }
+
+        let mut split_levels: Vec<Range<usize>> = Vec::new();
+        let mut current_level_start_ord = 0;
+        let mut current_level_max_docs = (splits[0].num_records * 3).max(self.min_level_num_docs);
+
+        for (split_ord, split) in splits.iter().enumerate() {
+            if split.num_records >= self.max_merge_docs {
+                break;
+            }
```
Comment (fmassot) [inline quickwit-indexing/src/merge_policy.rs:226]:
```suggestion
```

File: quickwit-indexing/src/merge_policy.rs (hunk 3 above)
Comment (fmassot) [reply quickwit-indexing/src/merge_policy.rs:167]:
```suggestion
```

File: quickwit-metastore/src/metastore.rs [hunk 6]
```diff
@@ -80,8 +80,9 @@ pub struct SplitMetadata {
     /// If a timestamp field is available, the min / max timestamp in the split.
     pub time_range: Option<RangeInclusive<i64>>,
 
-    /// Number of merges this segment has been subjected to during its lifetime.
-    pub generation: usize,
+    /// Encodes the type of demux operation the split has gone through.
+    /// Only splits who went through the same demux signature will be considered for demux.
+    pub demux_signature: String,
```
Comment (fmassot) [inline quickwit-metastore/src/metastore.rs:85]: Keep some meta information about previous operations sounds interesting. Of course we will have some logs but we can store at the split level operation count by operation type for example. Not super useful for now certainly.

**Review by fmassot — APPROVED** [review]

File: quickwit-indexing/src/merge_policy.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit-indexing/src/merge_policy.rs:116]: yes. renamed item / items

File: quickwit-indexing/src/actors/merge_planner.rs [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+        Ok(())
+    }
+}
+
+impl MergePlanner {
+    pub fn new(
+        merge_policy: Arc<dyn MergePolicy>,
+        merge_split_downloader_mailbox: Mailbox<MergeOperation>,
+    ) -> MergePlanner {
+        MergePlanner {
```
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit-indexing/src/actors/merge_planner.rs:75]: Yes it is done in the supervisor.

File: quickwit-indexing/src/merge_policy.rs [hunk 8]
```diff
@@ -0,0 +1,525 @@
...
+        if splits.is_empty() {
+            return Vec::new();
+        }
+
+        let mut split_levels: Vec<Range<usize>> = Vec::new();
+        let mut current_level_start_ord = 0;
+        let mut current_level_max_docs = (splits[0].num_records * 3).max(self.min_level_num_docs);
+
+        for (split_ord, split) in splits.iter().enumerate() {
+            if split.num_records >= self.max_merge_docs {
```
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit-indexing/src/merge_policy.rs:224]: oh yes you are right.

File: quickwit-metastore/src/metastore.rs (hunk 6 above)
Comment (fulmicoton) [reply quickwit-metastore/src/metastore.rs]: Removing demux_signature for the moment as it is premature.
It was not just a problem of observability. The idea was to use this information to identify which set of split can be demuxed together.

File: quickwit-indexing/Cargo.toml (hunk 1 above)
Comment (fulmicoton) [reply quickwit-indexing/Cargo.toml:46]: It's really good for that kind of job.

**Pull request merged**